src/shell/shell_cmds.cpp

```diff
@@ -465,26 +472,250 @@ static std::string to_search_pattern(const char *arg)
 	char buffer[CROSS_LEN];
 	pattern = ExpandDot(pattern.c_str(), buffer, sizeof(buffer));
 
-	// When there's no wildcard and target is a directory then search files
-	// inside the directory.
-	const char *p = pattern.c_str();
-	if (!strrchr(p, '*') && !strrchr(p, '?')) {
-		uint16_t attr = 0;
-		if (DOS_GetFileAttr(p, &attr) && (attr & DOS_ATTR_DIRECTORY))
-			pattern += "\\*.*";
+	return pattern;
+}
+
+uint32_t byte_count,file_count,dir_count;
+Bitu p_count;
+std::vector<std::string> dirs, adirs;
+
+static size_t GetPauseCount() {
+	return (real_readb(BIOSMEM_SEG,BIOSMEM_NB_ROWS) > 2u) ? (real_readb(BIOSMEM_SEG,BIOSMEM_NB_ROWS)) : 24u;
+}
+
+static bool dirPaused(DOS_Shell * shell, Bitu w_size, bool optP, bool optW) {
+	p_count+=optW?5:1;
+	if (optP && p_count%(GetPauseCount()*w_size)<1) {
+		shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+		Bit8u c;Bit16u n=1;
+		DOS_ReadFile(STDIN,&c,&n);
+		if (c==3) {shell->WriteOut("^C\r\n");return false;}
+		if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+		shell->WriteOut_NoParsing("\n");
 	}
+	return true;
+}
 
-	// If no extension, list all files.
-	// This makes patterns like foo* work.
-	if (!strrchr(pattern.c_str(), '.'))
-		pattern += ".*";
+static bool doDir(DOS_Shell * shell, char * args, DOS_DTA dta, char * numformat, Bitu w_size, bool optW, bool optZ, bool optS, bool optP, bool optB, bool optA, bool optAD, bool optAminusD, bool optAS, bool optAminusS, bool optAH, bool optAminusH, bool optAR, bool optAminusR, bool optAA, bool optAminusA, bool optO, bool optOG, bool optON, bool optOD, bool optOE, bool optOS, bool reverseSort) {
+	char path[LFN_NAMELENGTH+2];
+	char sargs[CROSS_LEN], largs[CROSS_LEN];
 
-	return pattern;
+	/* Make a full path in the args */
+	if (!DOS_Canonicalize(args,path)) {
+		shell->WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+		return true;
+	}
+	*(strrchr(path,'\\')+1)=0;
+	if (!DOS_GetSFNPath(path,sargs,false)) {
+		shell->WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+		return true;
+	}
+	if (!optB&&!optS) {
+		shell->WriteOut(MSG_Get("SHELL_CMD_DIR_INTRO"),uselfn&&!optZ&&DOS_GetSFNPath(path,largs,true)?largs:sargs);
+		if (optP) {
+			p_count+=optW?5:1;
+			if (p_count%(GetPauseCount()*w_size)<2) {
+				shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+				Bit8u c;Bit16u n=1;
+				DOS_ReadFile(STDIN,&c,&n);
+				if (c==3) {shell->WriteOut("^C\r\n");return false;}
+				if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+				shell->WriteOut_NoParsing("\n");
+			}
```

Comment:
Same comment as above.. and the code is looking pretty similar too. Can these two be lifted out into a function?

Reply:
@kcgen It can be lifted out into a function of course, but it will become a nested function (calling to the GetPauseCount() sub-function within this sub-function).